# Exploratory Data Analysis

In [13]:
import pandas as pd
import numpy as np

from pathlib import Path
from sklearn.impute import SimpleImputer

In [3]:
# Import training dataset
dataset = Path('../dataset')
df = pd.read_csv(dataset/"accidents_train.csv")
df.head()

Num      Time Day_of_week Age_band_of_driver Sex_of_driver  \
0    1  17:02:00      Monday              18-30          Male   
1    2  17:02:00      Monday              31-50          Male   
2    3  17:02:00      Monday              18-30          Male   
3    4   1:06:00      Sunday              18-30          Male   
4    5   1:06:00      Sunday              18-30          Male   

    Educational_level Vehicle_driver_relation Driving_experience  \
0   Above high school                Employee              1-2yr   
1  Junior high school                Employee         Above 10yr   
2  Junior high school                Employee              1-2yr   
3  Junior high school                Employee             5-10yr   
4  Junior high school                Employee              2-5yr   

       Type_of_vehicle Owner_of_vehicle  ... Vehicle_movement  \
0           Automobile            Owner  ...   Going straight   
1  Public (> 45 seats)            Owner  ...   Going straight   
2      Lorry (41?100Q)            Owner  ...   Going straight   
3  Public (> 45 seats)     Governmental  ...   Going straight   
4                  NaN            Owner  ...   Going straight   

    Casualty_class Sex_of_casualty Age_band_of_casualty Casualty_severity  \
0               na              na                   na                na   
1               na              na                   na                na   
2  Driver or rider            Male                31-50                 3   
3       Pedestrian          Female                18-30                 3   
4               na              na                   na                na   

  Work_of_casuality Fitness_of_casuality Pedestrian_movement  \
0               NaN                  NaN    Not a Pedestrian   
1               NaN                  NaN    Not a Pedestrian   
2            Driver                  NaN    Not a Pedestrian   
3            Driver               Normal    Not a Pedestrian   
4               NaN                  NaN    Not a Pedestrian   

            Cause_of_accident Accident_severity  
0             Moving Backward     Slight Injury  
1                  Overtaking     Slight Injury  
2   Changing lane to the left    Serious Injury  
3  Changing lane to the right     Slight Injury  
4                  Overtaking     Slight Injury  

[5 rows x 33 columns]

In [4]:
df.columns

Index(['Num', 'Time', 'Day_of_week', 'Age_band_of_driver', 'Sex_of_driver',
       'Educational_level', 'Vehicle_driver_relation', 'Driving_experience',
       'Type_of_vehicle', 'Owner_of_vehicle', 'Service_year_of_vehicle',
       'Defect_of_vehicle', 'Area_accident_occured', 'Lanes_or_Medians',
       'Road_allignment', 'Types_of_Junction', 'Road_surface_type',
       'Road_surface_conditions', 'Light_conditions', 'Weather_conditions',
       'Type_of_collision', 'Number_of_vehicles_involved',
       'Number_of_casualties', 'Vehicle_movement', 'Casualty_class',
       'Sex_of_casualty', 'Age_band_of_casualty', 'Casualty_severity',
       'Work_of_casuality', 'Fitness_of_casuality', 'Pedestrian_movement',
       'Cause_of_accident', 'Accident_severity'],
      dtype='object')

In [5]:
df.shape

(8210, 33)

Original dataset has 8210 rows  and 33 features. 
The target feature is "Accident_severity". 

In [6]:
target = 'Accident_severity'
my_list = df.columns.tolist()
set3 = my_list[21:33]
print(len(set3), set3)

12 ['Number_of_vehicles_involved', 'Number_of_casualties', 'Vehicle_movement', 'Casualty_class', 'Sex_of_casualty', 'Age_band_of_casualty', 'Casualty_severity', 'Work_of_casuality', 'Fitness_of_casuality', 'Pedestrian_movement', 'Cause_of_accident', 'Accident_severity']


## Dataset split into 3 sets of features for EDA by team

In [7]:
features = my_list[21:33]
df[features]

Number_of_vehicles_involved  Number_of_casualties Vehicle_movement  \
0                               2                     2   Going straight   
1                               2                     2   Going straight   
2                               2                     2   Going straight   
3                               2                     2   Going straight   
4                               2                     2   Going straight   
...                           ...                   ...              ...   
8205                            2                     1   Going straight   
8206                            2                     1          Unknown   
8207                            2                     1         Turnover   
8208                            2                     1      Getting off   
8209                            2                     1   Going straight   

       Casualty_class Sex_of_casualty Age_band_of_casualty Casualty_severity  \
0                  na              na                   na                na   
1                  na              na                   na                na   
2     Driver or rider            Male                31-50                 3   
3          Pedestrian          Female                18-30                 3   
4                  na              na                   na                na   
...               ...             ...                  ...               ...   
8205        Passenger            Male                31-50                 3   
8206        Passenger          Female              Over 51                 3   
8207        Passenger          Female                31-50                 3   
8208        Passenger          Female                31-50                 3   
8209        Passenger          Female              Over 51                 3   

     Work_of_casuality Fitness_of_casuality Pedestrian_movement  \
0                  NaN                  NaN    Not a Pedestrian   
1                  NaN                  NaN    Not a Pedestrian   
2               Driver                  NaN    Not a Pedestrian   
3               Driver               Normal    Not a Pedestrian   
4                  NaN                  NaN    Not a Pedestrian   
...                ...                  ...                 ...   
8205     Self-employed               Normal    Not a Pedestrian   
8206     Self-employed               Normal    Not a Pedestrian   
8207            Driver               Normal    Not a Pedestrian   
8208          Employee               Normal    Not a Pedestrian   
8209               NaN               Normal    Not a Pedestrian   

               Cause_of_accident Accident_severity  
0                Moving Backward     Slight Injury  
1                     Overtaking     Slight Injury  
2      Changing lane to the left    Serious Injury  
3     Changing lane to the right     Slight Injury  
4                     Overtaking     Slight Injury  
...                          ...               ...  
8205          Driving carelessly     Slight Injury  
8206   Changing lane to the left     Slight Injury  
8207             Moving Backward     Slight Injury  
8208                  Overtaking     Slight Injury  
8209   Changing lane to the left     Slight Injury  

[8210 rows x 12 columns]

In [1]:
# Import custom functions
%run ../custom/jc-functions.ipynb

## Data Understanding

In [8]:
# Set continuous and categorical features 
cont_feat = ['Number_of_vehicles_involved', 'Number_of_casualties']
cat_feat = ['Vehicle_movement', 'Casualty_class', 'Sex_of_casualty', 'Age_band_of_casualty', 'Casualty_severity', 'Work_of_casuality', 'Fitness_of_casuality', 'Pedestrian_movement', 'Cause_of_accident', 'Accident_severity']

#### Unique values for continuous features:

In [9]:
for feat in cont_feat:
    list = df[feat].unique()
    print(f"{feat}: There are {len(list)} unique items in this list. \n {list}")

Number_of_vehicles_involved: There are 6 unique items in this list. 
 [2 1 3 6 4 7]
Number_of_casualties: There are 8 unique items in this list. 
 [2 1 3 4 6 5 8 7]


#### Unique values for categorical features: 

In [10]:
for feat in cat_feat:
    list = df[feat].unique()
    print(f"{feat}: There are {len(list)} unique items in this list. \n {list}")

Vehicle_movement: There are 14 unique items in this list. 
 ['Going straight' 'U-Turn' 'Moving Backward' 'Turnover' 'Waiting to go'
 'Getting off' 'Reversing' 'Unknown' 'Parked' 'Stopping' 'Overtaking'
 'Other' 'Entering a junction' nan]
Casualty_class: There are 4 unique items in this list. 
 ['na' 'Driver or rider' 'Pedestrian' 'Passenger']
Sex_of_casualty: There are 3 unique items in this list. 
 ['na' 'Male' 'Female']
Age_band_of_casualty: There are 6 unique items in this list. 
 ['na' '31-50' '18-30' 'Under 18' 'Over 51' '5']
Casualty_severity: There are 4 unique items in this list. 
 ['na' '3' '2' '1']
Work_of_casuality: There are 8 unique items in this list. 
 [nan 'Driver' 'Other' 'Unemployed' 'Employee' 'Self-employed' 'Student'
 'Unknown']
Fitness_of_casuality: There are 6 unique items in this list. 
 [nan 'Normal' 'Deaf' 'Other' 'Blind' 'NormalNormal']
Pedestrian_movement: There are 9 unique items in this list. 
 ['Not a Pedestrian' "Crossing from driver's nearside"
 'Crossi

## Data Quality Report

### Continuous Features

In [11]:
data_quality_report_cont(df, cont_feat)

Data Quality for Continous Features
Total Features: 2
                       Feature  Count  Missing  % missing  Cardinality
0  Number_of_vehicles_involved   8210        0        0.0            6
1         Number_of_casualties   8210        0        0.0            8


Descriptive Stats
                              count  mean   std  min  25%  50%  75%  max
Number_of_vehicles_involved  8210.0  2.01  0.64  1.0  2.0  2.0  2.0  7.0
Number_of_casualties         8210.0  1.51  0.97  1.0  1.0  1.0  2.0  8.0


There are no missing values in the continuous variables. 
Therefore, no imputation or data cleaning required.

### Categorical Features

In [12]:
data_quality_report_cat(df, cat_feat)

Data Quality Report for Categorical Features
Stats
-----
                Feature  Count  Missing  % Missing  Cardinality
0      Vehicle_movement   8026      184       2.29           14
1        Casualty_class   8210        0       0.00            4
2       Sex_of_casualty   8210        0       0.00            3
3  Age_band_of_casualty   8210        0       0.00            6
4     Casualty_severity   8210        0       0.00            4
5     Work_of_casuality   6062     2148      35.43            8
6  Fitness_of_casuality   6440     1770      27.48            6
7   Pedestrian_movement   8210        0       0.00            9
8     Cause_of_accident   8210        0       0.00           20
9     Accident_severity   8210        0       0.00            3


Mode 1
------
                Feature            Mode 1  Mode 1 Freq.  Mode 1 %
0      Vehicle_movement    Going straight          5481     68.29
1        Casualty_class   Driver or rider          3201     38.99
2       Sex_of_casualty  

Three features have missing values. Less than 60% of the values are missing therefore missing values will be imputed using the mode.  

In [15]:
imputer = SimpleImputer(strategy='most_frequent')

col_to_impute = ['Vehicle_movement', 'Work_of_casuality', 'Fitness_of_casuality']

for col in col_to_impute:
    print(col)
    df[col] = imputer.fit_transform(df[[col]])


Vehicle_movement


ValueError: 2